### ISSUE TITLE
[Feature Request]: Metrics for Ratio of Sums, Mutual Information, and Pattern Matching

### ISSUE THREAD

**Issue by cobycloud** [swarmauri/swarmauri-sdk#491]
### Feature Name

Metrics for Ratio of Sums, Mutual Information, and Pattern Matching

### Feature Description

Ratio of Sums:

Explanation: Ratio of sums refers to the proportion between the sum of values in two or more datasets or columns. It provides insight into the relative magnitude of one set of values compared to another. For example, the ratio of total sales in one year to another year gives a sense of growth or decline.
Mutual Information:

Explanation: Mutual Information (MI) is a metric used to quantify the amount of information that two variables share. It measures the reduction in uncertainty about one variable given the knowledge of another. In other words, it shows how much knowing one variable tells you about the other. It’s widely used in feature selection and statistical analysis.
Pattern Match:

Explanation: Pattern matching involves searching for a sequence of characters or data that fit a predefined pattern. In the context of text or data analysis, it is used to identify if a string conforms to certain rules or patterns, such as email formats, specific keywords, or regular expressions.

### Motivation

Motivation for Ratio of Sums: The ratio of sums is a useful metric for comparing aggregate values across different periods, groups, or categories. It is widely used in business analytics to compare revenues, costs, or other metrics over time. For instance, a company might calculate the ratio of total sales this year to last year to understand growth rates. By using the ratio of sums, businesses can get a relative measure of performance, which is often more insightful than raw sums alone.

Motivation for Mutual Information: Mutual Information is a powerful tool for understanding relationships between variables, particularly in feature selection for machine learning. It helps in identifying which features provide the most information about the target variable, allowing developers to prioritize those variables in model development. Mutual Information is especially useful when working with non-linear relationships that traditional correlation measures might miss. In exploratory data analysis, MI provides insights into dependencies and reduces redundancy by highlighting features that are highly informative.

Motivation for Pattern Match: Pattern matching is essential in data validation and text processing tasks. For example, ensuring that all email addresses in a dataset conform to a valid format can prevent errors in communication systems or customer records. Pattern matching is widely used for validating inputs, cleaning data, and identifying specific sequences in textual data. By ensuring that data conforms to expected patterns, developers can improve the reliability of the system and prevent potential errors arising from malformed or incorrect inputs.

### Potential Solutions

```python
# Sample dataset for ratio of sums
data_ratio = pd.DataFrame({
    'Revenue_2023': [150, 200, 120, 300, 250],
    'Revenue_2022': [100, 150, 100, 250, 200]
})

# Calculating ratio of sums (Revenue 2023 to Revenue 2022)
sum_2023 = data_ratio['Revenue_2023'].sum()
sum_2022 = data_ratio['Revenue_2022'].sum()
ratio_of_sums = sum_2023 / sum_2022
ratio_of_sums

```

```python
from sklearn.feature_selection import mutual_info_classif
import numpy as np

# Sample dataset for mutual information
data_mi = pd.DataFrame({
    'Feature_1': [1, 0, 1, 1, 0, 0, 1, 0],
    'Feature_2': [1, 1, 0, 0, 1, 0, 1, 0]
})
target = np.array([1, 0, 1, 1, 0, 0, 1, 0])

# Calculating mutual information between Feature_1, Feature_2 and the target
mi = mutual_info_classif(data_mi, target)
mi


```

```python
# Sample dataset for pattern matching
data_pattern = pd.DataFrame({
    'Email': ['test@example.com', 'invalidemail', 'user@domain.com', 'admin@company']
})

# Pattern match for valid email format
pattern_match = data_pattern['Email'].str.contains(r'^[\w\.-]+@[\w\.-]+\.\w+$')
pattern_match


```

### Affected Areas

None

### Priority

Low
